comms/src/domain_connector.rs

```diff
@@ -0,0 +1,283 @@
+//  Copyright 2019 The Tari Project
+//
+//  Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+//  following conditions are met:
+//
+//  1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+//  disclaimer.
+//
+//  2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+//  following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+//  3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+//  products derived from this software without specific prior written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+//  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::{
+    connection::{zmq::ZmqEndpoint, Connection, ConnectionError, Direction, EstablishedConnection, ZmqContext},
+    message::{Frame, FrameSet},
+};
+use derive_error::Error;
+use serde::{Deserialize, Serialize};
+use std::{cmp, marker::PhantomData, time::Duration};
+use tari_utilities::message_format::{MessageFormat, MessageFormatError};
+
+#[derive(Debug, Error)]
+pub enum DomainConnectorError {
+    ConnectionError(ConnectionError),
+    #[error(no_from)]
+    ListenFailed(ConnectionError),
+    DeserializeFailed(MessageFormatError),
+    #[error(msg_embedded, non_std, no_from)]
+    ExtractFrameSetError(String),
+}
+
+/// # DomainConnector
+///
+/// The domain connector receives messages from an inbound address, and extracts and deserializes
+/// two frames into what the caller specifies.
+///
+/// This should be used by services/protocol managers to receive messages from the [InboundMessageService].
+pub struct DomainConnector<'de, H, T> {
+    connection: EstablishedConnection,
+    frame_extractor: Box<FrameExtractor<Error = DomainConnectorError>>,
+    _t: PhantomData<&'de (H, T)>,
+}
+
+impl<'de, H, T> DomainConnector<'de, H, T>
+where
+    T: Serialize + Deserialize<'de>,
+    H: Serialize + Deserialize<'de>,
```

Comment:
What are these generics for? With more reading I can see that the H is for Serializing/Deserializing the Header but what does T stand for? A comment on each here on this would help the readability of this a lot.

Reply:
Fair point will add a legend for the generics